File: graphs-grids.py
# bfs
"""
Explore a graph or tree level by level using a queue,
ideal for shortest path in **unweighted** graphs or grids and level‑order traversals.
Level order traversal
"""

from collections import deque

def bfs(start, neighbors_fn):
    """
    Generic BFS.
    - start: starting node/state.
    - neighbors_fn(node) -> iterable of neighbors.
    """
    visited = {start}
    queue = deque([(start, 0)])  # (node, distance)

    while queue:
        node, dist = queue.popleft()
        # TODO: process node (e.g., check goal, collect result)

        for nei in neighbors_fn(node):
            if nei not in visited:
                visited.add(nei)
                queue.append((nei, dist + 1))



  # dfs
"""
“All paths” problems, backtracking, or problems that naturally
decompose recursively on trees/graphs.
Tree problems asking to compute sums, depths, or collect root‑to‑leaf paths.
"""
  def dfs_paths(root):
    """
    Collect all root-to-leaf paths in a ***binary tree***.
    """
    res = []

    def dfs(node, path):
        if not node:
            return
        path.append(node.val)
        if not node.left and not node.right:
            res.append(list(path))
        else:
            dfs(node.left, path)
            dfs(node.right, path)
        path.pop()  # backtrack

    dfs(root, [])
    return res
  
def DFS(graph: dict[int,list[int]], start_vertex: int) -> :
  	visited = set(start_vertex)
  	stack = [start_vertex]
  	
  	while stack:
  		current_vertex = stack.pop()
  		visited.add(current_vertex)
  		for vertex in graph[current_vertex]:
  			if vertex not in visited:
  				stack.append(vertex)
  	return visited

  
# Binary tree traversals
"""
inorder: left - current node action - right
preorder: current node action - left - right
postorder: left - right - current node action
"""
def inorder(root):
    if not root:
        return
    inorder(root.left)
    # visit root.val
    inorder(root.right)

def preorder(root):
    if not root:
        return
    # visit root.val
    preorder(root.left)
    preorder(root.right)

def postorder(root):
    if not root:
        return
    postorder(root.left)
    postorder(root.right)
    # visit root.val

File: C++/graphs-grids.cpp
// Binary tree DFS and BFS
#include <queue>
#include <vector>
#include <iostream>

struct TreeNode {
    int val;
    TreeNode* left  = nullptr;
    TreeNode* right = nullptr;
    TreeNode(int v) : val(v) {}
};

// DFS — recursive inorder traversal
void inorder(TreeNode* root, std::vector<int>& result) {
    if (!root) return;
    inorder(root->left, result);
    result.push_back(root->val);
    inorder(root->right, result);
}

// BFS — level-order using a queue
std::vector<std::vector<int>> level_order(TreeNode* root) {
    std::vector<std::vector<int>> levels;
    if (!root) return levels;

    std::queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        int sz = q.size();
        std::vector<int> level;
        for (int i = 0; i < sz; ++i) {
            TreeNode* node = q.front(); q.pop();
            level.push_back(node->val);
            if (node->left)  q.push(node->left);
            if (node->right) q.push(node->right);
        }
        levels.push_back(level);
    }
    return levels;
}

int main() {
    //       1
    //      / \
    //     2   3
    TreeNode* root = new TreeNode(1);
    root->left  = new TreeNode(2);
    root->right = new TreeNode(3);

    std::vector<int> inorder_result;
    inorder(root, inorder_result);
    for (int x : inorder_result) std::cout << x << " ";  // 2 1 3
    std::cout << "\n";

    // Memory: in a real program use unique_ptr for tree nodes
    delete root->left; delete root->right; delete root;
    return 0;
}
